src/Features/LanguageServer/Protocol/Handler/Completion/CompletionListCache.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Completion
+{
+    /// <summary>
+    /// Caches completion lists in between calls to CompletionHandler and
+    /// CompletionResolveHandler. Used to avoid unnecessary recomputation.
+    /// </summary>
+    [Export(typeof(CompletionListCache)), Shared]
+    internal class CompletionListCache
+    {
+        /// <summary>
+        /// Maximum number of completion lists allowed in cache. Must be >= 1.
+        /// </summary>
+        private const int MaxCacheSize = 3;
```

Comment:
is this just for paranoia?  is there a case where we would ever be asked for anything but hte last computed completion list?  i don't mind this, i just want to understand if we have a legit case where we think we could have up to three outstanding completion sessions we were resolving against.

Reply:
Definitely more of a precaution - I think you're right in that we shouldn't normally be asked for anything but the last computed completion list, but I'm not sure if LSP will always abide by that or if things may get out-of-sync somehow.

Reply:
sgtm.  it won't hurt to have a few extra cache entries :)